### ISSUE TITLE
CPU implementation?

### ISSUE THREAD

**Issue by PetrochukM** [openai/blocksparse#19]
Hi!

Are you aware of a CPU implementation for this technique? Trying to reimplement this paper: https://arxiv.org/pdf/1802.08435.pdf

Thanks!

**Comment by ThomasHagebols** [comment]
[Disclaimer] I'm not the project maintainer

What do you want to achieve? Do you only want to do inference on the cpu of do you need to train as well? For now I assume you need both inference and training on the cpu

I'm not aware of any blocksparse cpu implementations. Some time ago I built a deep learning framework from scratch using the bsr_matrix data structure from scipy (with some custom Cython code for calculating gradients). Though this is not a strategy I can recommend

I think writing a custom PyTorch layer using bsr_matrix and Cython might work okish for research purposes. Though I would not recommend this for production.

For information on writing custom layers in pytorch see:
https://pytorch.org/tutorials/beginner/pytorch_with_examples.html#pytorch-custom-nn-modules

The forward pass can be executed using default bsr_matrix operations. For the backward pass you need a custom Cython operation. You only want to calculate the gradient of the active weights in the weight matrix. If you do a naive dot product of the activations and the grad_output you get a dense matrix, which is not what you want.

**Comment by PetrochukM** [comment]
Hi There!

Looking to only do inference. 

Worked on this for a couple of days by now. Turns out SciPy is not the best choice. SciPy got me no speed up despite having 95% + sparsity. We moved on and used https://software.intel.com/en-us/mkl and got a 7x speed up.

It'd be nice to do training as well, however, that's not a bottleneck at the moment.

Thanks!
